client/extensions/woocommerce/app/promotions/fields/applies-to-single-product-field.js

```diff
@@ -0,0 +1,167 @@
+/**
+ * External dependencies
+ */
+import React from 'react';
+import PropTypes from 'prop-types';
+import { connect } from 'react-redux';
+import classNames from 'classnames';
+
+/**
+ * Internal dependencies
+ */
+import FormField from './form-field';
+import FormLabel from 'components/forms/form-label';
+import FormRadio from 'components/forms/form-radio';
+import { getProductCategories } from 'woocommerce/state/sites/product-categories/selectors';
+import { getPromotionableProducts } from 'woocommerce/state/selectors/promotions';
+import { getSelectedSiteWithFallback } from 'woocommerce/state/sites/selectors';
+import Search from 'components/search';
+
+function productContainsString( product, textString ) {
+	const matchString = textString.trim().toLowerCase();
+
+	if ( -1 < product.name.toLocaleLowerCase().indexOf( matchString ) ) {
+		// found in product name
+		return true;
+	}
+
+	return false;
+}
+
+function isProductSelected( appliesTo, productId ) {
+	if ( ! appliesTo || ! appliesTo.productIds ) {
+		return false;
+	}
+
+	return ( -1 !== appliesTo.productIds.indexOf( productId ) );
+}
+
+class AppliesToSingleProductField extends React.Component {
+	static propTypes = {
+		fieldName: PropTypes.string.isRequired,
+		labelText: PropTypes.string.isRequired,
+		explanationText: PropTypes.string,
+		isRequired: PropTypes.bool.isRequired,
+		value: PropTypes.object,
+		edit: PropTypes.func.isRequired,
+		products: PropTypes.array,
+		productCategories: PropTypes.array,
+	};
+
+	constructor( props ) {
+		super( props );
+		this.state = {
+			searchFilter: '',
+		};
+	}
+
+	getFilteredProducts() {
+		const { value, products } = this.props;
+		const { searchFilter } = this.state;
+
+		if ( 0 === searchFilter.length ) {
+			return products;
+		}
+
+		return products && products.filter(
+			( product ) => {
+				return productContainsString( product, searchFilter ) ||
+					isProductSelected( value, product.id );
```

Comment:
I like how we're keeping the selected product - the UX is less confusing that way

Reply:
Yeah, it seemed the least confusing way to do it, but I think it's still just a little confusing.